lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -220,3 +694,263 @@ class ConnectorSourceNotify[In: Any val]
+          s.last_acked = s.last_seen
+        end
+      end
+    end
+
+  fun ref barrier_complete(checkpoint_id: CheckpointId) =>
+    // update barrier state and check checkpoint_id matches our local knowledge
+    _barrier_ongoing = false
+    Invariant(checkpoint_id == _barrier_checkpoint_id)
+
+    if _session_active then
+      _process_acks_for_active()
+
+      // process acks for EOS/_pending_close streams
+      _process_acks_for_pending_close()
+      // process any stream relinquish requests
+      _relinquish_streams()
+    end
+
+  /////////////////////////
+  // Local state management
+  /////////////////////////
+  fun ref _process_acks_for_active() =>
+    for (stream_id, s) in _active_streams.pairs() do
+      _pending_acks.push((stream_id, s.last_acked))
+    end
+    if (_fsm_state is _ProtoFsmStreaming) or
+       (_fsm_state is _ProtoFsmShrinking) then
+      _send_acks()
+    end
+
+  fun ref _process_acks_for_pending_close() =>
+    for (stream_id, s) in _pending_close.pairs() do
+      ifdef "trace" then
+        @printf[I32]("%s ::: Processing ack for %s at %s\n".cstring(),
+          WallClock.seconds().string().cstring(),
+          stream_id.string().cstring(), s.last_acked.string().cstring())
+      end
+      _pending_acks.push((stream_id, s.last_acked))
+      _pending_relinquish.push(StreamTuple(s.id, s.name, s.last_acked))
+    end
+    // clear _pending_close
+    _pending_close.clear()
+    if (_fsm_state is _ProtoFsmStreaming) or
+       (_fsm_state is _ProtoFsmShrinking) then
+      _send_acks()
+    end
+
+  fun ref _relinquish_streams() =>
+    let streams = recover iso Array[StreamTuple] end
+    for s in _pending_relinquish.values() do
+      streams.push(s)
+    end
+    _pending_relinquish.clear()
+    if streams.size() > 0 then
+      @printf[I32]("ConnectorSource relinquishing %s streams\n".cstring(),
+        _pending_relinquish.size().string().cstring())
+      _listener.streams_relinquish(consume streams)
+    end
+
+  fun ref _clear_streams() =>
+    _pending_notify.clear()
+    _active_streams.clear()
+    _pending_close.clear()
+    _pending_relinquish.clear()
+
+  fun ref _clear_and_relinquish_all() =>
+    for s_map in [_active_streams ; _pending_close].values() do
+      for s in s_map.values() do
+        // This should never happen, but you never know
+        Invariant(s.last_seen == s.last_acked)
+        _pending_relinquish.push(StreamTuple(s.id, s.name, s.last_acked))
+      end
+    end
+    _relinquish_streams()
+    _clear_streams()
+
+  /////////
+  // Notify
+  /////////
+
+  fun ref _process_notify(source: ConnectorSource[In] ref, stream_id: StreamId,
+    stream_name: String, point_of_reference: PointOfReference)
+  =>
+    // add to _pending_notify
+    _pending_notify.set(stream_id)
+
+    // create a promise for handling result
+    let promise = Promise[NotifyResult[In]]
+    promise.next[None](recover NotifyResultFulfill[In](stream_id, _session_id) end)
```

Comment:
Why can't this be:
```
promise.next[None](_connector_source~stream_notify_result(_session_id))
```
and then ConnectorSource.stream_notify_result takes an object that has the success and stream info on it?

Reply:
@JONBRWN Can you address this?